### PR TITLE
Fix setting a role with non-existing controller or action

### PR DESCRIPTION
**Pull request by Yoann-Abbes**
## What does this PR do ?

This PR checks controller and actions of the role which is being set and throw an error if the controller or the action does not exist

Before this PR:
When you set a right on a non-existing controller or action, Kuzzle allow you to do that without any warning.

Now:
We throw an error when user try to set rights to non-existing controller or action of the Kuzzle API.

Concerning plugin API, it could be convenient to define roles on plugin controller that will be deployed in a near futur. 
To do that we have now a `force` option that allows users to write invalid plugin role.

In the case `force` is set to `true` we notify the user that his role is created but there might be some error.
In the case `force` is not set or set to `false` we throw a proper error.

Also, a sanity check is added at kuzzle startup to notify every invalid existing plugin role.

Jira [ticket](https://jira.kaliop.net/browse/KZL-1356)

### How should this be manually tested?

  - Step 1 :
  - Step 2 :
  - Step 3 :  
  ...

### Other changes

`searchRole` returned an array randomly sorted
Now it returns a sorted array by `_id`

### CODE REVIEW

File: lib/core/models/repositories/roleRepository.js [hunk 1]
```diff
@@ -181,25 +195,80 @@ class RoleRepository extends Repository {
...
+                'unknown_controller_action',
+                role._id,
+                action,
+                controller);
+            }
+          });
+        });
+      }
+      else {
+        if (!this.kuzzle.funnel.isNativeController(
```
Comment (Aschen) [inline lib/core/models/repositories/roleRepository.js]: I think this method signature should be modified to take only the controller name as an argument.
Also, if the role define rights for a plugin controller, this method will throw (eg: `s3/upload` ). You should not check anything if the controller isn't a native one

File: lib/core/models/repositories/roleRepository.js [hunk 2]
```diff
@@ -181,25 +195,80 @@ class RoleRepository extends Repository {
...
+              'role',
+              'unknown_controller_action',
+              role._id,
+              action,
+              roleController);
+          }
+        });
+      }
+    });
+    return Promise.resolve();
```
Comment (Aschen) [inline lib/core/models/repositories/roleRepository.js]: This methods does not need to returns a promise

File: lib/core/models/repositories/roleRepository.js [hunk 3]
```diff
@@ -181,25 +195,80 @@ class RoleRepository extends Repository {
...
       });
   }
 
+  /**
+   * Given a Role object, checks if its controllers and actions exist.
+   *
+   * @param {Role} role
+   * @returns Promise
+   */
+  checkRoleControllersAndActions(role) {
```
Comment (Aschen) [inline lib/core/models/repositories/roleRepository.js]: Nit/proposal: could be renamed because it should checks only for native action existence

File: lib/core/models/repositories/roleRepository.js [hunk 4]
```diff
@@ -212,18 +281,21 @@ class RoleRepository extends Repository {
...
         }
 
         return this.deleteFromDatabase(role._id, options)
-          .then(deleteResponse => {
-            this.roles.delete(role._id);
-            this.kuzzle.emit('core:roleRepository:delete', {_id: role._id});
-            return deleteResponse;
-          });
+          .then(
+            deleteResponse => {
```
Comment (Aschen) [inline lib/core/models/repositories/roleRepository.js]: Why did you change the indent here?

File: lib/config/error-codes/security.json [hunk 5]
```diff
@@ -176,6 +176,18 @@
           "message": "Missing or malformed \"test\" attribute  for role %s (%s) : %s",
           "class": "BadRequestError",
           "deprecated": true
+        },
+        "unknown_controller": {
```
Comment (Aschen) [inline lib/config/error-codes/security.json]: This error should not exists since we cannot tell if the user misspell a native controller name or just wanted to set a plugin controller right

File: lib/core/models/repositories/roleRepository.js [hunk 6]
```diff
@@ -181,25 +195,118 @@ class RoleRepository extends Repository {
...
-
-        return this.persistToDatabase(role, options);
+        this.kuzzle.emit('core:roleRepository:save', {
+          _id: role._id,
+          controllers: role.controllers
+        });
       })
+      .then(() => this.checkRoleNativeRights(role))
+      .then(err => {
+        err = this.checkRolePluginsRights(role);
```
Comment (kuzzle) [inline lib/core/models/repositories/roleRepository.js]: MINOR Introduce a new variable instead of reusing the parameter "err". [rule](https://sonarqube.kaliop.net/coding_rules#rule_key=javascript%3AS1226)

File: lib/core/models/repositories/roleRepository.js [hunk 7]
```diff
@@ -181,25 +195,118 @@ class RoleRepository extends Repository {
...
+   */
+  checkRolePluginsRights(role) {
+    const roleRightsError = errorsManager.wrap('security', 'role');
+    const errors = [];
+    Object.keys(role.controllers).forEach(roleController => {
+      if (roleController.split('/').length === 1) {
+        // Meaning this is not a plugin controller. So we skip it.
+        return;
+      }
+      const pluginName = roleController.split('/')[0];
```
Comment (kuzzle) [inline lib/core/models/repositories/roleRepository.js]: MINOR Use destructuring syntax for these assignments from "roleController.split('/')". [rule](https://sonarqube.kaliop.net/coding_rules#rule_key=javascript%3AS3514)

**Comment by kuzzle** [comment]
SonarQube analysis reported 2 issues
* ![MINOR][MINOR] 2 minor

Watch the comments in this conversation to review them.

[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

**Comment by Yoann-Abbes** [comment]
Functional tests are OK with this branch of the SDK JS : https://github.com/kuzzleio/sdk-javascript/pull/476

File: lib/core/models/repositories/roleRepository.js [hunk 8]
```diff
@@ -181,25 +194,130 @@ class RoleRepository extends Repository {
...
+    Object.keys(role.controllers).forEach(roleController => {
+      if (roleController !== '*' && !this.kuzzle.funnel.controllers.has(roleController)) {
+        if (roleController.split('/').length > 1) {
+          // Meaning this is a plugin controller. So we skip it.
+          return;
+        }
+        roleRightsError.throw('unknown_controller', role._id, roleController);
+      }
+      if (roleController === '*') {
+        this.kuzzle.funnel.controllers.forEach(controller => {
```
Comment (Aschen) [inline lib/core/models/repositories/roleRepository.js]: I don't see the point to check if an action really exists on all controllers mostly because I don't see the usecase here. I think it's better to forbid this kind of rights.

```js
controllers: {
  '*': {
    actions: {
      create: true // all controllers must have a "create" action otherwise this role is invalid
    }
  }
}
```

We should only accept the a wildcard controller with a wildcard action

File: lib/core/models/repositories/roleRepository.js [hunk 9]
```diff
@@ -181,25 +194,130 @@ class RoleRepository extends Repository {
...
+  }
+
+  /**
+   * Given a Role object, checks if its controllers and actions exist in plugins.
+   *
+   * @param {Role} role
+   * @param {Force} force
+   */
+  checkRolePluginsRights(role, force = false) {
+    const roleRightsError = errorsManager.wrap('security', 'role');
```
Comment (Aschen) [inline lib/core/models/repositories/roleRepository.js]: You should declare this variable in the global scope instead of re-declaring it each time the method is called

File: lib/core/models/repositories/roleRepository.js [hunk 10]
```diff
@@ -181,25 +194,130 @@ class RoleRepository extends Repository {
...
+    Object.keys(role.controllers).forEach(roleController => {
+      if (roleController.split('/').length === 1) {
```
Comment (Aschen) [inline lib/core/models/repositories/roleRepository.js]: Much faster test
```suggestion
      if (roleController.indexOf('/') === -1) {
```

File: lib/core/models/repositories/roleRepository.js [hunk 11]
```diff
@@ -181,25 +194,130 @@ class RoleRepository extends Repository {
...
+      const pluginController = roleController.split('/')[1];
+      const availablePlugins = Object.keys(this.kuzzle.pluginsManager.plugins);
+      if (!availablePlugins.includes(pluginName)) {
+        if (!force) {
+          roleRightsError.throw('unavailable_plugin', role._id, pluginName);
+        }
+        this.kuzzle.log.warn(`[!WARNING!] Trying to set role "${role._id}" with plugin "${pluginName}" which doesn't exist or is not enabled.`);
+        return;
+      }
+      const roleActions = Object.keys(role.controllers[`${pluginName}/${pluginController}`].actions);
```
Comment (Aschen) [inline lib/core/models/repositories/roleRepository.js]: Why reconstruct a string since you can use the `roleController` variable?

File: lib/core/models/repositories/roleRepository.js [hunk 12]
```diff
@@ -181,25 +194,130 @@ class RoleRepository extends Repository {
...
+    });
+  }
+
+  /**
+   * Given a Role object, checks if its controllers and actions exist in plugins.
+   *
+   * @param {Role} role
+   * @param {Force} force
+   */
+  checkRolePluginsRights(role, force = false) {
```
Comment (Aschen) [inline lib/core/models/repositories/roleRepository.js]: Use named parameters syntax instead for readability
```suggestion
  checkRolePluginsRights(role, { force=false } = {}) {
```
So you can call it like this 
```
checkRolePluginsRights(role, { force: true })
checkRolePluginsRights(role) // force: false by default
```

File: lib/core/models/repositories/roleRepository.js [hunk 13]
```diff
@@ -181,25 +194,130 @@ class RoleRepository extends Repository {
...
+
+  /**
+   * Fetching roles and check for each of them for invalid plugin rights.
+   * If there are some, Kuzzle will log a warning.
+   */
+
+  checkAllRolePluginsRights() {
+    return this.search({}, {})
+      .then(roles => roles.hits.forEach(role => {
+        this.checkRolePluginsRights(role, true);
```
Comment (Aschen) [inline lib/core/models/repositories/roleRepository.js]: This check should be done in the `RoleRepository.init` method.

File: lib/kuzzle.js [hunk 14]
```diff
@@ -199,6 +199,7 @@ class Kuzzle extends EventEmitter {
         return this.validation.curateSpecification();
       })
       .then(() => this.entryPoints.init())
+      .then(() => this.repositories.role.checkAllRolePluginsRights())
```
Comment (Aschen) [inline lib/kuzzle.js]: This check should be handled by the `RoleRepository.init` method which is called by the `Repositories.init` method (and this last method is called in Kuzzle initialization)

File: features-sdk/step_definitions/security-steps.js [hunk 15]
```diff
@@ -14,10 +14,91 @@ Given('I create a profile {string} with the following policies:', async function
...
 });
 
-Given('I create a role {string} with the following API rights:', async function (roleId, dataTable) {
+Given(/I (can not )?create a role "(.*?)" with the following API rights:/, async function (not, roleId, dataTable) {
 
   const controllers = this.parseObject(dataTable);
-  this.props.result = await this.sdk.security.createRole(roleId, { controllers }, { refresh: 'wait_for' });
+
+  try {
+    this.props.result = await this.sdk.security.createRole(roleId, { controllers }, { refresh: 'wait_for' });
```
[the remark this answers is not in the document]
Comment (Yoann-Abbes) [reply features-sdk/step_definitions/security-steps.js]: I think maybe we need it because after `create` or `update` role there are some find and search in steps

[the remark this answers is not in the document]
Comment (Aschen) [reply features-sdk/step_definitions/security-steps.js]: We definitively miss a `collection:refreshInternal` method to refresh internal collection when needed. In the meantime we need to keep the `refresh` yes

File: doc/2/api/controllers/security/create-or-replace-role/index.md [hunk 16]
```diff
@@ -63,6 +63,8 @@ Body:
 
 - `refresh`: if set to `wait_for`, Kuzzle will not respond until the created/replaced role is indexed (default: `"wait_for"`)
 
+- `force`: if set to `true`, Kuzzle will force the creation of a role even if it contains invalid plugin API rights. However, even if you use this option, you can't set invalid rights about the native API.
```
Comment (scottinet) [inline doc/2/api/controllers/security/create-or-replace-role/index.md]:
```suggestion
- `force`: if set to `true`, creates or replaces the role even if it gives access to non-existent plugins API routes.
```

File: doc/2/api/controllers/security/create-role/index.md [hunk 17]
```diff
@@ -63,6 +63,8 @@ Body:
 
 - `refresh`: if set to `wait_for`, Kuzzle will not respond until the created role is indexed (default: `"wait_for"`)
 
+- `force`: if set to `true`, Kuzzle will force the creation of a role even if it contains invalid plugin API rights. However, even if you use this option, you can't set invalid rights about the native API.
```
Comment (scottinet) [inline doc/2/api/controllers/security/create-role/index.md]:
```suggestion
- `force`: if set to `true`, creates the role even if it gives access to non-existent plugins API routes.
```

File: doc/2/api/controllers/security/update-role/index.md [hunk 18]
```diff
@@ -65,6 +65,8 @@ Body:
 
 - `refresh`: if set to `wait_for`, Kuzzle will not respond until the role changes are indexed (default: `"wait_for"`)
 
+- `force`: if set to `true`, Kuzzle will force the creation of a role even if it contains invalid plugin API rights. However, even if you use this option, you can't set invalid rights about the native API.
```
Comment (scottinet) [inline doc/2/api/controllers/security/update-role/index.md]:
```suggestion
- `force`: if set to `true`, updates the role even if it gives access to non-existent plugins API routes.
```

File: doc/2/api/essentials/errors/codes/index.md [hunk 19]
```diff
@@ -475,6 +475,11 @@ order: 500
 | security.role.cannot_delete<br/><pre>0x07050003</pre> | [BadRequestError](/core/2/api/essentials/errors/handling#badrequesterror) <pre>(400)</pre> | Attempted to delete a base role (anonymous, default, admin) |
 | security.role.in_use<br/><pre>0x07050004</pre> | [PreconditionError](/core/2/api/essentials/errors/handling#preconditionerror) <pre>(412)</pre> | A role still assigned to profiles cannot be deleted |
 | security.role.uninitialized<br/><pre>0x07050005</pre> | [InternalError](/core/2/api/essentials/errors/handling#internalerror) <pre>(500)</pre> | Attempted to access to an unitialized Role object |
+| security.role.unknown_controller<br/><pre>0x07050009</pre> | [BadRequestError](/core/2/api/essentials/errors/handling#badrequesterror) <pre>(400)</pre> | Trying to set a role with a non-existing controller |
+| security.role.unknown_controller_action<br/><pre>0x0705000a</pre> | [BadRequestError](/core/2/api/essentials/errors/handling#badrequesterror) <pre>(400)</pre> | Trying to set a role with a non-existing controller action |
```
Comment (scottinet) [inline doc/2/api/essentials/errors/codes/index.md]: Should be kept short: `security.role.unknown_action`

File: doc/2/api/essentials/errors/codes/index.md [hunk 20]
```diff
@@ -475,6 +475,11 @@ order: 500
...
+| security.role.unavailable_plugin<br/><pre>0x0705000b</pre> | [BadRequestError](/core/2/api/essentials/errors/handling#badrequesterror) <pre>(400)</pre> | Trying to set a role with a non-existing plugin |
```
Comment (scottinet) [inline doc/2/api/essentials/errors/codes/index.md]: Should be `security.role.unknown_plugin` to be consistent with other errors

File: doc/2/api/essentials/errors/codes/index.md [hunk 21]
```diff
@@ -475,6 +475,11 @@ order: 500
...
+| security.role.unavailable_controller_plugin<br/><pre>0x0705000c</pre> | [BadRequestError](/core/2/api/essentials/errors/handling#badrequesterror) <pre>(400)</pre> | Trying to set a role with a non-existing plugin controller |
```
Comment (scottinet) [inline doc/2/api/essentials/errors/codes/index.md]: Redundant with `security.role.unknown_controller`

File: doc/2/api/essentials/errors/codes/index.md [hunk 22]
```diff
@@ -475,6 +475,11 @@ order: 500
...
+| security.role.unavailable_action_plugin<br/><pre>0x0705000d</pre> | [BadRequestError](/core/2/api/essentials/errors/handling#badrequesterror) <pre>(400)</pre> | Trying to set a role with a non-existing plugin controller action |
```
Comment (scottinet) [inline doc/2/api/essentials/errors/codes/index.md]: Redundant with `security.role.unknown_action`

File: lib/config/error-codes/security.json [hunk 23]
```diff
@@ -176,6 +176,36 @@
...
```
[the remark this answers is not in the document]
Comment (scottinet) [reply lib/config/error-codes/security.json]: I find this error ok (but not others, see my comments), because either way, the provided controller is effectively unknown to the role validator.

I would add a didYouMean hint to the message however

File: lib/core/models/repositories/roleRepository.js [hunk 24]
```diff
@@ -181,25 +194,130 @@ class RoleRepository extends Repository {
    * @returns Promise
    */
   validateAndSaveRole (role, options) {
-    return role.validateDefinition()
+
+    return role
```
Comment (scottinet) [inline lib/core/models/repositories/roleRepository.js]: asyncify this method instead of extending the promise chain

File: lib/core/models/repositories/roleRepository.js [hunk 25]
```diff
@@ -181,25 +194,130 @@ class RoleRepository extends Repository {
...
+  /**
+   * Given a Role object, checks if its controllers and actions exist.
+   *
+   * @param {Role} role
+   */
+  checkRoleNativeRights(role) {
+    const roleRightsError = errorsManager.wrap('security', 'role');
+    Object.keys(role.controllers).forEach(roleController => {
+      if (roleController !== '*' && !this.kuzzle.funnel.controllers.has(roleController)) {
+        if (roleController.split('/').length > 1) {
```
[the remark this answers is not in the document]
Comment (scottinet) [reply lib/core/models/repositories/roleRepository.js]: ... or just use `this.kuzzle.funnel.isNativeController(<string>)`

[the remark this answers is not in the document]
Comment (scottinet) [reply lib/core/models/repositories/roleRepository.js]: Also I find this code brittle: it's not the responsibility of the role repository to know the difference between a native controller and a plugin one.

Instead, I would either use `this.kuzzle.funnel.getController(<Request>)`, or create a new dedicated funnel getter named `this.kuzzle.funnel.exists(controller, action)`. That new getter code would take its code after the existing getController function.

Either way, the funnel is the API entrypoint, and it alone should know whether a controller/action pair exists or not: keep a clean separation of concern

File: lib/core/models/repositories/roleRepository.js [hunk 26]
```diff
@@ -181,25 +194,130 @@ class RoleRepository extends Repository {
...
+        this.kuzzle.log.warn(`[!WARNING!] Trying to set role "${role._id}" with plugin "${pluginName}" which doesn't exist or is not enabled.`);
+        return;
+      }
+      const roleActions = Object.keys(role.controllers[`${pluginName}/${pluginController}`].actions);
+      const pluginControllers = Object.keys(this.kuzzle.pluginsManager.plugins[pluginName].object.controllers);
+      if (pluginController !== '*' && !pluginControllers.includes(pluginController)) {
+        if (!force) {
+          roleRightsError.throw('unavailable_controller_plugin', role._id, pluginController, pluginName);
+        }
+        this.kuzzle.log.warn(`[!WARNING!] Trying to set role "${role._id}" with a non-existing controller "${pluginController}" in the plugin "${pluginName}".`);
```
Comment (scottinet) [inline lib/core/models/repositories/roleRepository.js]: Misleading: it's not "trying" since it allows that creation, hence the warning.

```suggestion
        this.kuzzle.log.warn(`[!WARNING!] Role "${role._id}" gives access to the non-existing controller "${pluginController}" in the plugin "${pluginName}".`);
```

File: lib/core/models/repositories/roleRepository.js [hunk 27]
```diff
@@ -181,25 +194,130 @@ class RoleRepository extends Repository {
...
+        return;
+      }
+      const pluginActions = [];
+      this.kuzzle.pluginsManager.plugins[pluginName].object.routes.forEach(route => {
```
Comment (scottinet) [inline lib/core/models/repositories/roleRepository.js]: This violates the single responsibility principle.
Create dedicated getters in the plugins manager module instead of digging in its internal structures: it's the manager's responsibility to store and manager plugin routes, controllers and actions.

for instance, a `pluginsManager.getRoutes(<plugin name>)` method returning an array of routes

that way, you could write:
```
const pluginActions = this.kuzzle.pluginsManager.getRoutes(pluginName).map(r => r.action);
```

File: lib/core/models/repositories/roleRepository.js [hunk 28]
```diff
@@ -181,25 +194,130 @@ class RoleRepository extends Repository {
...
+    });
+    return;
+  }
+
+  /**
+   * Fetching roles and check for each of them for invalid plugin rights.
+   * If there are some, Kuzzle will log a warning.
+   */
+
+  checkAllRolePluginsRights() {
```
Comment (scottinet) [inline lib/core/models/repositories/roleRepository.js]: use async/await instead of a promise chain

File: lib/core/models/repositories/roleRepository.js [hunk 29]
```diff
@@ -181,25 +194,130 @@ class RoleRepository extends Repository {
...
+      const pluginActions = [];
+      this.kuzzle.pluginsManager.plugins[pluginName].object.routes.forEach(route => {
+        pluginActions.push(route.action);
+      });
+      roleActions.forEach(action => {
+        if (action !== '*' && !pluginActions.includes(action)) {
+          if (!force) {
+            roleRightsError.throw('unavailable_action_plugin', role._id, action, pluginController, pluginName);
+          }
+          this.kuzzle.log.warn(`[!WARNING!] Trying to set role "${role._id}" with a non-existing action "${action}" for the controller "${pluginController}" in the plugin "${pluginName}".`);
```
Comment (scottinet) [inline lib/core/models/repositories/roleRepository.js]:
```suggestion
          this.kuzzle.log.warn(`[!WARNING!] The role "${role._id}" gives access to the non-existing action "${action}" for the controller "${pluginController}" in the plugin "${pluginName}".`);
```

File: lib/kuzzle.js (hunk 14 above)
Comment (Yoann-Abbes) [reply lib/kuzzle.js]: How I am supposed to do it since this check use the `pluginsManager` which at this state is not initialized ?

File: lib/core/models/repositories/roleRepository.js [hunk 30]
```diff
@@ -180,24 +196,131 @@ class RoleRepository extends Repository {
...
+  /**
+   * Given a Role object, checks if its controllers and actions exist in plugins.
+   *
+   * @param {Role} role
+   * @param {Force} force
+   */
+  checkRolePluginsRights(role, { force=false } = {}) {
+    Object.keys(role.controllers).forEach(roleController => {
+      if ( !(roleController !== '*'
+        && !this.kuzzle.funnel.isNativeController(roleController))
```
[the remark this answers is not in the document]
Comment (Yoann-Abbes) [reply lib/core/models/repositories/roleRepository.js]: In fact, im checking if its a native one because if its not, then its a plugin one.

File: features-sdk/SecurityController.feature [hunk 31]
```diff
@@ -132,3 +132,96 @@ Feature: Security Controller
     Then I delete the user "test-user"
     And I delete the profile "test-profile"
     And I delete the role "test-role"
+
+  @security
+  Scenario: Create a role with invalid API rights
+    When I can not "create" a role "test-role" with the following API rights:
```
[the remark this answers is not in the document]
Comment (Yoann-Abbes) [reply features-sdk/SecurityController.feature]: That is why just above there is
```
 Then I should receive an error matching:
    | id | "security.role.unknown_controller" |
```

**Comment by Yoann-Abbes** [comment]
Reminder: All functional test are OK with the `7-dev` branch of the SDK JS. (the `force` option into the sdk is not released yet)

File: lib/core/models/repositories/roleRepository.js (hunk 30 above)
[the remark this answers is not in the document]
Comment (xbill82) [reply lib/core/models/repositories/roleRepository.js]: Ok, sorry, didn't see the exclamation mark that negates the whole predicate. So this condition is OK!

File: lib/core/plugins/manager.js [hunk 32]
```diff
@@ -311,9 +311,10 @@ class PluginsManager {
     }
 
     if (preparedPipes.length === 1) {
-      callback(null, ...data);
-    } else {
-      this.pipeRunner.run(preparedPipes, callback);
+      callback.call(callbackContext, null, ...data);
```
Comment (xbill82) [inline lib/core/plugins/manager.js]: Wow... This is a big change, isn't it? What issue does this fix?

File: features-sdk/SecurityController.feature (hunk 31 above)
[the remark this answers is not in the document]
Comment (xbill82) [reply features-sdk/SecurityController.feature]: Indeed. I am not questioning the `Then` step, but the `When`, which, IMHO, should state `When I try to create`, rather than `When I can not create`.

File: lib/core/plugins/manager.js (hunk 32 above)
Comment (Yoann-Abbes) [reply lib/core/plugins/manager.js]: I believe its https://github.com/kuzzleio/kuzzle/pull/1557/

File: lib/core/models/repositories/roleRepository.js [hunk 33]
```diff
@@ -180,24 +182,126 @@ class RoleRepository extends Repository {
...
-        return updatedRole;
+  /**
+   * Given a Role object, checks if its controllers and actions exist in plugins.
+   *
+   * @param {Role} role
+   * @param {Force} force
+   */
+  checkRolePluginsRights(role, { force=false } = {}) {
+    Object.keys(role.controllers).forEach(roleController => {
+      if ( !(roleController !== '*'
```
Comment (scottinet) [inline lib/core/models/repositories/roleRepository.js]: (nitpicking) the following would be much simpler to read (and to maintain):

```js
      if ( roleController === '*' 
        || this.kuzzle.funnel.isNativeController(roleController)
      ) {
```

File: lib/core/plugins/manager.js [hunk 34]
```diff
@@ -92,16 +92,77 @@ class PluginsManager {
...
+    return Object.keys(this.plugins).includes(pluginName);
+  }
+
+  /**
+   * Tell if the controller is a plugin controller or not
+   *
+   * @param {String} controller
+   * @returns {Boolean}
+   */
+  isPluginController (controller) {
```
Comment (scottinet) [inline lib/core/plugins/manager.js]: redundant with the function `isController` above
